### PR TITLE
feat: new `no-new-native-nonconstructor` rule

### LINKED ISSUE #16322: New Rule: `no-new-bigint`
### Rule details

Restricts calling `BigInt` with `new` like [`no-new-symbol`](https://eslint.org/docs/latest/rules/no-new-symbol)

### Related ECMAScript feature

[BigInt](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt)

### What type of rule is this?

Warns about a potential problem

### Example code

```js
// valid
const a = BigInt(9007199254740991);

// invalid
const b = new BigInt(9007199254740991);
```

### Why should this rule be in the core instead of a plugin?

`BigInt` is a core feature of ECMAScript, also ESLint have [`no-new-symbol`](https://eslint.org/docs/latest/rules/no-new-symbol) already.

### Participation

- [X] I am willing to submit a pull request to implement this rule.

### Additional comments

_No response_

### PR DESCRIPTION
**Pull request by sosukesuzuki**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[x] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

Fixes #16322

Adds new rule `no-new-nonconstructor`.

#### Is there anything you'd like reviewers to focus on?

question: Should we deprecate `no-new-symbol` in this PR? or in other PR?

### CODE REVIEW

File: lib/rules/no-new-nonconstructor.js
Comment (sosukesuzuki) [inline lib/rules/no-new-nonconstructor.js:3]: Just a question, most of the code for this rule is the same as `no-new-symbol`. Who should we mention as a an author?

File: lib/rules/no-new-nonconstructor.js
Comment (sosukesuzuki) [inline lib/rules/no-new-nonconstructor.js:25]: Should be recommended?

**Comment by ljharb** [comment]
The naming of this rule doesn't just imply inclusion of primitive constructors that throw, but also all HTML element constructors, as well as `%TypedArray%` (altho that one's hard to lint for) - but also *any function* that's not newable, which includes arrows, async, generators, async generators, concise object methods, and class methods.

Is that broad scope intended?

**Comment by sosukesuzuki** [comment]
> Is that broad scope intended?

No. Only built-in functions such as `Symbol` and `BigInt` are intended. We need a better name. Like `no-new-builtin-nonconstructor`?

**Comment by ljharb** [comment]
In the future, when there's other constructors you can't `new`, would they be included by implication? Is it just primitive constructors? What if there's a future `Decimal`, or `Record` or `Tuple`, that have the same semantics?

If not, it seems like a `no-new-bigint` rule would be better. If so, then it's important to define exactly what the criteria is, and the name should be informed by that.

**Comment by mdjermanovic** [comment]
The criteria is: ES global variable that is a function (noncallable globals are covered by the `no-obj-calls` rule) and starts with an uppercase letter ( = looks like a constructor) but can't be used with `new`. At the moment, that would be `Symbol` and `BigInt`. In the future, new ES globals that satisfy the criteria would be added. It wouldn't be restricted to globals that relate to primitive types.

**Comment by ljharb** [comment]
and "can't be used with `new`" means that you get an exception when you invoke it with `new`, i gather?

what about `HTMLElement` and all the other HTML element global constructors that can't be used with `new`?

**Comment by mdjermanovic** [comment]
> and "can't be used with `new`" means that you get an exception when you invoke it with `new`, i gather?

Yes, exactly.

> what about `HTMLElement` and all the other HTML element global constructors that can't be used with `new`?

This rule checks only globals from the ECMAScript specification, currently Symbol and BigInt, similar to the [no-obj-calls](https://eslint.org/docs/latest/rules/no-obj-calls) rule which checks only Math, JSON, Reflect and Atomics.

**Comment by sosukesuzuki** [comment]
I don't think `no-new-noconstructor` is a bad name. It is similar to `no-obj-calls`.

I think it is sufficient to specify [the criteria](https://github.com/eslint/eslint/pull/16368#issuecomment-1265039565) in the documentation.

File: lib/rules/no-new-nonconstructor.js
Comment (nzakas) [reply lib/rules/no-new-nonconstructor.js:3]: Yes, that would be nice.

File: lib/rules/no-new-nonconstructor.js
Comment (nzakas) [reply lib/rules/no-new-nonconstructor.js:25]: No. We can't set to recommended until we do a major release. We can switch it during the next major release.

**Comment by nzakas** [comment]
We could go with `no-new-native-nonconstructor` instead?

File: docs/src/rules/no-new-native-nonconstructor.md
Comment (mdjermanovic) [inline docs/src/rules/no-new-native-nonconstructor.md:58]: This shouldn't be correct code for this rule?

File: docs/src/rules/no-new-native-nonconstructor.md
Comment (mdjermanovic) [inline docs/src/rules/no-new-native-nonconstructor.md:54]:
```suggestion
/*eslint no-new-native-nonconstructor: "error"*/
```

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

File: lib/rules/no-new-native-nonconstructor.js
Comment (mdjermanovic) [inline lib/rules/no-new-native-nonconstructor.js:24]:
```suggestion
            description: "Disallow `new` operators with global non-constructor functions",
```

File: docs/src/rules/no-new-native-nonconstructor.md
Comment (sosukesuzuki) [reply docs/src/rules/no-new-native-nonconstructor.md:9]: Sounds good

File: docs/src/_data/further_reading_links.json
Comment (mdjermanovic) [inline docs/src/_data/further_reading_links.json:702]:
```suggestion
    "https://tc39.es/ecma262/#sec-symbol-constructor": {
```

This should fix the docs site build.

File: docs/src/_data/further_reading_links.json
Comment (mdjermanovic) [inline docs/src/_data/further_reading_links.json:709]:
```suggestion
    "https://tc39.es/ecma262/#sec-bigint-constructor": {
```

Comment (sosukesuzuki) [reply docs/src/_data/further_reading_links.json]: @mdjermanovic Thank you! I've fixed via running `docs:update-links`

File: docs/src/rules/no-new-native-nonconstructor.md
Comment (mdjermanovic) [inline docs/src/rules/no-new-native-nonconstructor.md:6]:
```suggestion
- no-obj-calls
```

File: docs/src/rules/no-new-native-nonconstructor.md
Comment (mdjermanovic) [inline docs/src/rules/no-new-native-nonconstructor.md:21]:
```suggestion
var bar = new BigInt(9007199254740991);
```

File: docs/src/rules/no-new-native-nonconstructor.md
Comment (mdjermanovic) [inline docs/src/rules/no-new-native-nonconstructor.md:54]:
```suggestion
/*eslint no-new-native-nonconstructor: "error"*/
```

File: lib/rules/no-new-native-nonconstructor.js
Comment (mdjermanovic) [inline lib/rules/no-new-native-nonconstructor.js:24]:
```suggestion
            description: "Disallow `new` operators with global non-constructor functions",
```

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

I'll leave this open for a few days in case someone else wants to review it before merging.

**Review by nzakas — APPROVED** [review]
LGTM. Thanks!

**Comment by nzakas** [comment]
@mdjermanovic because this is meant to replace no-new-symbol, so we also want to deprecate no-new-symbol or otherwise indicate the relationship between these two rules somehow?

**Comment by mdjermanovic** [comment]
We should deprecate `no-new-symbol`, but since it's in `eslint:recommended` then maybe better to deprecate it in the next major version?

**Review by snitin315 — APPROVED** [review]
LGTM, thanks!

**Comment by snitin315** [comment]
> We should deprecate no-new-symbol, but since it's in eslint:recommended then maybe better to deprecate it in the next major version?

I feel we can deprecate in minor versions, and in the major version remove `no-new-symbol` in favor of `no-new-native-nonconstructor`. Deprecations are meant to warn users about a change beforehand.

**Review by btmills — APPROVED** [review]
LGTM, thanks!

I just resolved a merge conflict in the further reading links. Another entry had been added since this branch was last updated to `main`, so I just needed to add an extra `},` line. I'll merge this once CI passes.

@snitin315 in most cases we could deprecate a rule in a minor release, but we're unable to add `no-new-native-nonconstructor` to `eslint:recommended` until the next major release, so resolving the deprecation warning would require users to manually override `eslint:recommended` even if they previously had no customizations. I will, however, call out the replacement in the release notes.

**Pull request merged**
